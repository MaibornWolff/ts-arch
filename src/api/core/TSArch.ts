import { EntryPipe } from './pipe/EntryPipe'

export class TSArch {
  static defineThat(): EntryPipe {
    return new EntryPipe()
  }
}
