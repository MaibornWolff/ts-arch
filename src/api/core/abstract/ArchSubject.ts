export abstract class ArchSubject {
  constructor(private name: string)

  public getName(): string {
    return this.name
  }
}
