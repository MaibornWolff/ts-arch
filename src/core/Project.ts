import { File } from "./noun/File"
import { Result } from "./Result"
import { Checkable } from "./checks/Checkable"

export class Project {
	constructor(private files: File[] = []) {}

	public addFile(file: File) {
		this.files.push(file)
	}

	public getFiles(): File[] {
		return this.files
	}

	check(ruleToMatch: Checkable): Result {
		return ruleToMatch.check(this.files)
	}
}
