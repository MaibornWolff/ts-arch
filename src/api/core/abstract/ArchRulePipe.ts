import { ArchSubject } from './ArchSubject'

export abstract class ArchRulePipe {
  constructor(private input: ArchRulePipe) {}

  public filterSubjects(subjects: ArchSubject[]): ArchSubject[] {
    return this.input ? this.input.filterSubjects(subjects) : subjects
  }

  public hasNotModifier(): boolean {
    return this.input ? this.input.hasNotModifier() : false
  }
}
