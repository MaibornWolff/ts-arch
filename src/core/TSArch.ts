import { RuleBuilder } from "./builder/RuleBuilder"
import { SubjectFilterStarter } from "./lang/SubjectFilterStarter"
import { Project } from "./Project"
import { ProjectParser } from "./parser/ProjectParser"
import { TSArchConfig } from "./TSArchConfig";

export class TSArch {
	public static config: TSArchConfig = {
		ignore: {
			declarations: true,
			nodeModules: true,
			js: false
		}
	}

	static defineThat(): SubjectFilterStarter {
		return new RuleBuilder(TSArch.config.ignore)
	}

	static async parseTypescriptProject(path: string): Promise<Project> {
		const project = await ProjectParser.parse(path, TSArch.config.ignore)
		return project
	}
}
