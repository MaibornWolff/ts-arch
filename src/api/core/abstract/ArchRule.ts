import { ArchSubject } from './ArchSubject'
import { ArchRulePipe } from './ArchRulePipe'

export abstract class ArchRule {
  constructor(private input: ArchRulePipe) {}
  abstract checkCondition(subjects: ArchSubject[]): boolean
  public check(subjects: ArchSubject[]): boolean {
    return this.input.hasNotModifier()
      ? !this.checkCondition(subjects)
      : this.checkCondition(subjects)
  }
}
