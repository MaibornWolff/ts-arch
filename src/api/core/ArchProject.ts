import { ArchSubject } from "./abstract/ArchSubject"

export class ArchProject {
	constructor(private subjects: ArchSubject[] = []) {}

	public addSubject(subject: ArchSubject) {
		this.subjects.push(subject)
	}

	public getSubjects(): ArchSubject[] {
		return this.subjects
	}
}
