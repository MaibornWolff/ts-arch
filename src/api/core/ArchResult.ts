import { ArchSubject } from "./abstract/ArchSubject"

export class ArchResult {
	private results: Array<{ subject: ArchSubject; info: string; pass: boolean }> = []
	private hasForcePassed = false
	private hasForceFailed = false

	public forcePass() {
		this.hasForcePassed = true
		this.hasForceFailed = false
	}

	public forceFail() {
		this.hasForcePassed = false
		this.hasForceFailed = true
	}

	public addEntry(subject: ArchSubject, pass: boolean, info: string) {
		this.results.push({ subject: subject, info: info, pass: pass })
	}

	public getEntries(): Array<{ subject: ArchSubject; info: string; pass: boolean }> {
		return this.results
	}

	public hasRulePassed(): boolean {
		if (this.hasForcePassed) {
			return true
		}
		if (this.hasForceFailed) {
			return false
		}
		let result = true
		this.results.forEach(r => {
			if (!r.pass) {
				result = false
			}
		})
		return result
	}
}
