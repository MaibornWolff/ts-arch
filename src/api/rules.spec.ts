import { TSArch } from './core/TSArch'
import { ArchSubject } from './core/abstract/ArchSubject'
describe('Rules', () => {
  let subjects

  beforeEach(() => {
    subjects = [
      new ArchSubject('DogService'),
      new ArchSubject('DogController'),
      new ArchSubject('DogModel'),
      new ArchSubject('CatController'),
      new ArchSubject('CatFactory')
    ]
  })

  it('all subjects should have no subjects is false', () => {
    const rule = TSArch.defineThat()
      .allSubjects()
      .should()
      .not()
      .haveSubjects()
    expect(rule.check(subjects)).toBeFalsy()
  })

  it('all subjects should have subjects is true', () => {
    const rule = TSArch.defineThat()
      .allSubjects()
      .should()
      .haveSubjects()
    expect(rule.check(subjects)).toBeTruthy()
  })

  it('all subjects with name matching x should have subjects is false', () => {
    const rule = TSArch.defineThat()
      .allSubjects()
      .withNameMatching(/x/)
      .should()
      .haveSubjects()
    expect(rule.check(subjects)).toBeFalsy()
  })

  it('all subjects with name matching .* should have subjects is true', () => {
    const rule = TSArch.defineThat()
      .allSubjects()
      .withNameMatching(/.*/)
      .should()
      .haveSubjects()
    expect(rule.check(subjects)).toBeTruthy()
  })

  it('should find all controllers', () => {
    const pipe = TSArch.defineThat()
      .allSubjects()
      .withNameMatching(/Controller/)
    expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
  })

  it('should find all Dog stuff', () => {
    const pipe = TSArch.defineThat()
      .allSubjects()
      .withNameMatching(/Dog/)
    expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
  })

  it('should find all Services by suffix', () => {
    const pipe = TSArch.defineThat()
      .allSubjects()
      .withNameSuffix('Service') //TODO this is the wrong result
    expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
  })

  it('should find all Dog things by prefix', () => {
    const pipe = TSArch.defineThat()
      .allSubjects()
      .withNamePrefix('Dog') //TODO this is the wrong result
    expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
  })

  it('should not find all Dog things by suffix', () => {
    const pipe = TSArch.defineThat()
      .allSubjects()
      .withNameSuffix('Dog') //TODO this is the wrong result
    expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
  })

  it('should not find all Services by prefix', () => {
    const pipe = TSArch.defineThat()
      .allSubjects()
      .withNamePrefix('Service') //TODO this is the wrong result
    expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
  })
})
