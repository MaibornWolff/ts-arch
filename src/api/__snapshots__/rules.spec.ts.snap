// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rules should find all Dog stuff 1`] = `
Array [
  FileSubject {
    "name": "DogService",
    "path": "C://project/src/service",
  },
  FileSubject {
    "name": "DogController",
    "path": "C://project/src/controller",
  },
  FileSubject {
    "name": "DogModel",
    "path": "C://project/src/model",
  },
]
`;

exports[`Rules should find all Dog things by prefix 1`] = `
Array [
  FileSubject {
    "name": "DogService",
    "path": "C://project/src/service",
  },
  FileSubject {
    "name": "DogController",
    "path": "C://project/src/controller",
  },
  FileSubject {
    "name": "DogModel",
    "path": "C://project/src/model",
  },
]
`;

exports[`Rules should find all Services by file suffix 1`] = `
Array [
  FileSubject {
    "name": "DogService",
    "path": "C://project/src/service",
  },
]
`;

exports[`Rules should find all Services by matching path 1`] = `
Array [
  FileSubject {
    "name": "DogService",
    "path": "C://project/src/service",
  },
]
`;

exports[`Rules should find all Services by suffix 1`] = `
Array [
  FileSubject {
    "name": "DogService",
    "path": "C://project/src/service",
  },
]
`;

exports[`Rules should find all controllers 1`] = `
Array [
  FileSubject {
    "name": "DogController",
    "path": "C://project/src/controller",
  },
  FileSubject {
    "name": "CatController",
    "path": "C://project/src/controller",
  },
]
`;

exports[`Rules should not find all Dog things by suffix 1`] = `Array []`;

exports[`Rules should not find all Services by prefix 1`] = `Array []`;
